### ISSUE TITLE
xtest 600x causes memory leak followed by memory corruption

### ISSUE THREAD

**Issue by lorc** [OP-TEE/optee_os#4326]
HW: RCAR Gen3.
Setup without virtualization enabled.

I ran ``for x in `seq 50`; do xtest 600; done``

At some iteration this happened:

```

* regression_6009 Test TEE Internal API Persistent Object Enumeration Functions                                                                                                                
o regression_6009.1 Storage id: 00000001                                                                                                                                                       
/home/lorc/work/optee_test/host/xtest/regression_6000.c:1043: fs_create(&sess, file_02, sizeof(file_02), 0x00000002, 0, data_01, sizeof(data_01), &obj2, storage_id) has an unexpected value: 0
xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0x0 = TEEC_SUCCESS                                                                                                                                regression_6009.1 FAILED                                                                                                                                                                     
o regression_6009.2 Storage id: 80000000                                                                                                                                                       
/home/lorc/work/optee_test/host/xtest/regression_6000.c:1029: fs_create(&sess, file_00, sizeof(file_00), 0x00000002, 0, data_01, sizeof(data_01), &obj0, storage_id) has an unexpected value: 0
xffff0003 = TEEC_ERROR_ACCESS_CONFLICT, expected 0x0 = TEEC_SUCCESS                                                                                                                              regression_6009.2 FAILED                                                                                                                                                                     
  regression_6009 FAILED    

```

After this every 60xx test fails with the `TEEC_ERROR_ACCESS_CONFLICT`:

```
* regression_6001 Test TEE_CreatePersistentObject                
o regression_6001.1 Storage id: 00000001/home/lorc/work/optee_test/host/xtest/regression_6000.c:686: fs_create(&sess, file_00, sizeof(file_00), 0x00000002 | 0x00000004, 0, data_00, sizeof(data_00), &obj, storage_id) has an unexpect
ed value: 0xffff0003 = TEEC_ERROR_ACCESS_CONFLICT, expected 0x0 = TEEC_SUCCESS                                                                                                                   regression_6001.1 FAILED                                                                     
o regression_6001.2 Storage id: 80000000
/home/lorc/work/optee_test/host/xtest/regression_6000.c:686: fs_create(&sess, file_00, sizeof(file_00), 0x00000002 | 0x00000004, 0, data_00, sizeof(data_00), &obj, storage_id) has an unexpect
ed value: 0xffff0003 = TEEC_ERROR_ACCESS_CONFLICT, expected 0x0 = TEEC_SUCCESS
  regression_6001.2 FAILED                   
  regression_6001 FAILED   
```              

Apart from that some other test are failing with `TEEC_ERROR_OUT_OF_MEMORY`:

```
* regression_1008 TEE internal client API                                                                                                                                                      
o regression_1008.1 Invoke command                                                                                                                                                             
/home/lorc/work/optee_test/host/xtest/regression_1000.c:716: TEEC_InvokeCommand(&session, 7, ((void *)0), &ret_orig) has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0
x0 = TEEC_SUCCESS               
  regression_1008.1 FAILED                                                                     
o regression_1008.2 Invoke command with timeout                                                                                                                                                
  regression_1008.2 OK                                                                         
o regression_1008.3 Create session fail                                                        
n = 0                                                                                                                                                                                          
  regression_1008.3 OK                 
o regression_1008.4 Load corrupt TA                                                                                                                                                            
/home/lorc/work/optee_test/host/xtest/regression_1000.c:659: res has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0xffff000f = TEEC_ERROR_SECURITY                     
/home/lorc/work/optee_test/host/xtest/regression_1000.c:688: load_corrupt_ta(c, __builtin_offsetof (struct shdr, img_size), 1) has an unexpected value: 0x0 = false, expected 0x1 = true       /home/lorc/work/optee_test/host/xtest/regression_1000.c:659: res has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0xffff000f = TEEC_ERROR_SECURITY                     
/home/lorc/work/optee_test/host/xtest/regression_1000.c:700: load_corrupt_ta(c, 3000, 1) has an unexpected value: 0x0 = false, expected 0x1 = true                                             
/home/lorc/work/optee_test/host/xtest/regression_1000.c:659: res has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0xffff000f = TEEC_ERROR_SECURITY                     
/home/lorc/work/optee_test/host/xtest/regression_1000.c:701: load_corrupt_ta(c, 8000, 1) has an unexpected value: 0x0 = false, expected 0x1 = true                                             
  regression_1008.4 FAILED                                                                                                                                                                     
  regression_1008 FAILED  
```

**Comment by jenswi-linaro** [comment]
This is probably fixed by https://github.com/OP-TEE/optee_os/pull/4329

**Comment by lorc** [comment]
I can confirm that #4329 fixes the issue with the memory leak.

But I am worried about the storage corruption. I'm not very familiar with storage subsystem, so I want to clarify: Is this the inherent problem of the current design, or some bug? Looks like storage operations are not atomic and failure amidst an operation can corrupt the whole storage.

**Comment by jenswi-linaro** [comment]
I expect that storage errors are due to the test cases not cleaning up on failure.
Especially with the`TEEC_ERROR_ACCESS_CONFLICT` value.

**Comment by lorc** [comment]
Ah, I see... Then #4330 is not an actual issue.

Is there a way to force xtest to do the clean up?

**Comment by jenswi-linaro** [comment]
If xtest crashes or is killed it will not be able to do that. Perhaps we could add an option to xtest which could try to clean secure storage, either by nuking it entirely or just for a specific TA. In either case it will be a bit dangerous and must not be possible to do on a real product.

**Comment by lorc** [comment]
Yes, I fixed this by nuking all the data. But I'm worried that this approach can hide potential issues. 

Anyways, it seems that this is the different story, so this issue can be closed with #4329 merged.
